### PR TITLE
Docker images

### LINKED ISSUE #1013: Feedback on "Building Docker Images for .NET Core Applications"
The following are some review comments I have on the [Building Docker Images for .NET Core Applications](https://docs.microsoft.com/en-us/dotnet/articles/core/docker/building-net-docker-images) documentation.

1. _Docker Image Optimizations - Production_ This section references fast deploy and start time.  Another important aspect of production images that isn't mentioned is around the required space to run a container and achieving high disk density across the running containers.  In the case of .NET, by having the shared framework in the base image, you amortize its size across the number of containers running on the same host because you are able to share the layer.
2. _Docker image variations-sdk_ This section references has the wrong tag for the sdk image - microsoft/dotnet-preview2-sdk.  It should be microsoft/dotnet:1.0.0-preview2-sdk note the version and the tag separator.
3. _Alternative images - onbuild_ This section should reference the versions onbuild tag to be consistent with the other image descriptions - e.g. microsoft/dotnet:1.0.0-preview2-onbuild vs microsoft/dotnet:onbuild.  The versioned onbuild images should also appear in the block that lists the various images and their sizes.
4. _Prerequisites_ I see Docker for Windows or Docker for Mac is required.  What is the limitation on using Linux?
5. _Creating the Web API application_ I was a little surprised to see "before we containerize the application, first run the application locally".  I was expecting that the development image would be used.  The whole purpose of having a development image is so that you do not have configure a machine.  You should be able to use a predefined image (with maybe some additions to add some preferred or optional tools).

@spboyer, @dleeapho

### PR DESCRIPTION
**Pull request by spboyer**
## Summary

Fixes #1013 
## Suggested Reviewers

@BillWagner @MichaelSimons  @stevelasker @glennc

### CODE REVIEW

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.io.unmanagedmemoryaccessor?branch=pr-en-us-1014)

File: docs/core/docker/building-net-docker-images.md
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: I was expecting to still see the onbuild tag listed separately to be consistent with core, core-deps and latest.

File: docs/core/docker/building-net-docker-images.md
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: This doesn't address my concern about the Production documentation.  The documentation states the reason for a small image is that it can be spun up quickly.  It is more than that IMO.  You also want a small image so that you can have many containers running on a given machine.  That being said, you can't simply look at the image size to determine this, you have to look at the size of the layers that can be shared.  In other words are you achieving high disk density with your containers?  With the dotnet:core image, users can hopefully achieve relatively high disk density because the shared framework layer can be shared across all containers.

**Comment by MichaelSimons** [comment]
> - [Docker for Windows or Docker for Mac](http://www.docker.com/products/docker) from Docker to run your Docker containers locally 

I realize that the majority of dev platforms will be OSX and Windows but I was just surprised to see the instructions exclude Linux.  What is the harm in simply stating Docker needs to be installed instead of explicitly stating Docker for Windows or Docker for Mac?

---
Refers to: docs/core/docker/building-net-docker-images.md:78 in 27f6f82. [](commit_id = 27f6f8292ab7d0c6b9e1a01862bb03d1a406b2f9, deletion_comment = False)

**Comment by spboyer** [comment]
@MichaelSimons no harm, made the change for "Docker"

File: docs/core/docker/building-net-docker-images.md
Comment (spboyer) [inline docs/core/docker/building-net-docker-images.md]: I'll defer to @glennc  and @stevelasker on how they would like to present the statement.  The statement is true across containers in general not, just .NET core

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.runtime.interopservices.osplatform?branch=pr-en-us-1014)

File: docs/core/docker/building-net-docker-images.md
Comment (glennc) [inline docs/core/docker/building-net-docker-images.md]: This is a difficult topic to tackle in a doc like this. In reality, the importance of image size is a combination of:

Surface area
Download size
Virtual Size (including layering)

And when you start discussing density you very quickly veer into runtime memory efficiency vs disk density.

We don't want to turn this into a massive document trying to teach you everything about Docker, layer sharing, and the security benefits of surface area and complexity. It is designed to answer the question of when I would want to use one image as a base over another. Right?

Comment (BillWagner) [inline docs/core/docker/building-net-docker-images.md]: How about stating something like this:

There are many Docker images you can build from because different scenarios will benefit from different configurations. Images with fewer dependencies may be more beneficial for scenarios where all your containers are running the same application: The smaller base footprint is more beneficial than the amount of shared dependencies. Scenarios involving many applications deployed on many containers will benefit from using a base image with more dependencies. Sharing dependencies is more beneficial than the smaller footprint for the image. 

To determine the best base image for your scenario, you'll need to measure using different base images.  That's described (in a document that isn't written yet, but should be on the backlog :smile: )

Comment (SteveLasker) [inline docs/core/docker/building-net-docker-images.md]: There's lots of additional info that could be written. What's the goal of this page? Our goal was to explain the different images. The primary images developers should consider, and we offer some additional images, for non-mainline scenarios. The more we write, the less value it provides as customers will get lost in the details. That said, having more detail on advanced image layering isn't a bad idea. I just don't believe this is the place to do it. If there's a docker page we can point to, ...for more information on docker image layering, see here.

**Comment by spboyer** [comment]
Where do you all want this to land?  Leave as is or merge this?

**Review by BillWagner — APPROVED** [review]
I'm ready to merge if @MichaelSimons and @SteveLasker are ready.

**Comment by MichaelSimons** [comment]
Go ahead and merge.

**Comment by SteveLasker** [comment]
@sboyer, can we get the new adpnetcore image added as well? Our new 0.40 tools target it

**Comment by spboyer** [comment]
@SteveLasker yes we can add aspnetcore.

**Pull request merged**
